open_spiel/games/dou_dizhu/dou_dizhu_utils.cc

```diff
@@ -0,0 +1,860 @@
+    {hand[params.chain_head + i] = used_rank[params.chain_head + i] = 3;}
+  const int kicker_steps = params.kicker_id;
+  int count = 0;
+  bool found = dfs_airplane_kicker(params.chain_length, 0, kicker_steps, 
+        count, kNumRanks-1, used_rank.begin(), 
+          hand.begin(), params.kicker_type);
+  SPIEL_CHECK_TRUE(found);
+  return hand;
+}
+
+
+
+// for aiplane combination, we have to specify the chain head
+// to resolve ambiguity such as 333444555666
+int AirplaneCombHandToActionId(const std::array<int, kNumRanks>& hand, 
+        int chain_head, KickerType kicker_type){
+  int chain_length = 0;
+  bool chain_begun = false;
+  std::vector<int> kickers;
+  for(int rank = 0; rank < kNumRanks; ++rank){
+    SPIEL_CHECK_LT(hand[rank], kNumSuits);
+    if(!hand[rank]) continue;
+    if(!chain_begun && rank != chain_head) {
+      if(kicker_type == kSolo) 
+        for(int i = 0; i < hand[rank]; ++i) kickers.push_back(rank);
+      else{
+        SPIEL_CHECK_EQ(hand[rank], 2); kickers.push_back(rank);
+      }
+    }else if(rank == chain_head) {
+      SPIEL_CHECK_EQ(hand[rank], 3);
+      chain_begun = true; chain_length++;
+    } else if(chain_begun && hand[rank] == 3) chain_length++;
+    else if(chain_begun && hand[rank] != 3){
+      chain_begun = false;
+      if(kicker_type == kSolo)
+        for(int i = 0; i < hand[rank]; ++i) kickers.push_back(rank);
+      else{
+        SPIEL_CHECK_EQ(hand[rank], 2); kickers.push_back(rank);
+      }
+    }
+  }
+
+  // handle case where 333444555666 and chain_head=3
+  // in this case, the above linear scan algorithm will view 3-4-5-6 as the chain
+  // where 6s should be the kickers
+  if(chain_length - 1 == static_cast<int>(kickers.size()) + 3){
+    chain_length--;
+    for(int i = 0; i < 3; ++i) kickers.push_back(chain_head+chain_length);
+  }
+  SPIEL_CHECK_EQ(chain_length, static_cast<int>(kickers.size()));
+
+  if(chain_head + chain_length - 1 >= kNumRanks - 3) SpielFatalError("2s, Joker cannot be in a chain");
+  int action_base;
+  if(kicker_type == kSolo)
+    action_base = GetAirplaneSoloActionBase(chain_length) + chain_head * GetNumKickersAirplaneSoloComb(chain_length); 
+  else action_base = GetAirplanePairActionBase(chain_length) + chain_head * GetNumKickersAirplanePairComb(chain_length);
+
+  int count = 0;
+  std::array<int, kNumRanks> used_rank{};
+  for(int i = 0; i < chain_length; ++i) used_rank[chain_head+i] = 3;
+
+
+  std::array<int, kNumRanks> hand_copy(hand);
+  bool found = dfs_airplane_kicker(chain_length, 0, -1, count, kNumRanks-1, 
+    used_rank.begin(), hand_copy.begin(), kicker_type);
+  SPIEL_CHECK_TRUE(found);
+  
+  return action_base + count;
+}
+
+
+
+
+// a dfs backtrack algorithm that found the action ids of all possible airplane combination
+// the action ids are stored in action_ids reference
+void dfs_add_all_airplane_kickers(int chain_head, int chain_length, int depth, int max_search_rank,
+      int* used_rank, const int* ans_hand, 
```

Comment:
Similarly, here, I think that you should make this signature:

```
absl::Span<int> used_rank, absl::Span<const int> ans_hand, std::vector<Action>* action_ids
```

(and call `absl::MakeSpan(used_rank)` on the calling code).